watcher/tool/watcher/autoscale.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/monitoring-app/watcher/lib/constants"
+	"github.com/gravitational/monitoring-app/watcher/lib/kubernetes"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/client/versioned/typed/monitoring/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+type autoscaleConfig struct {
+	// nodes is the Nodes API client.
+	nodes v1.NodeInterface
+	// alertmanagers is the Alertmanagers CRD API client.
+	alertmanagers monitoringv1.AlertmanagerInterface
+	// prometheuses is the Prometheuses CRD API client.
+	prometheuses monitoringv1.PrometheusInterface
+	// interval is the reconciliation interval.
+	interval time.Duration
+}
+
+func (c *autoscaleConfig) checkAndSetDefaults() error {
+	if c.nodes == nil {
+		return trace.BadParameter("missing Nodes client")
+	}
+	if c.alertmanagers == nil {
+		return trace.BadParameter("missing Alertmanagers client")
+	}
+	if c.prometheuses == nil {
+		return trace.BadParameter("missing Prometheuses client")
+	}
+	if c.interval == 0 {
+		c.interval = time.Minute
+	}
+	return nil
+}
+
+func runAutoscale(ctx context.Context, config autoscaleConfig) error {
+	err := config.checkAndSetDefaults()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	ticker := time.NewTicker(config.interval)
+	defer ticker.Stop()
+
+	log.Info("Starting autoscaler.")
+
+	for {
+		select {
+		case <-ticker.C:
+			nodes, err := getMasterNodes(config.nodes)
+			if err != nil {
+				log.WithError(err).Error("Failed to query nodes.")
+				continue
+			}
+			err = reconcileAlertmanager(config.alertmanagers, nodes)
```

Comment:
nit: given an unlikely broken configuration when alertmanager updates would always fail, this will not be able to scale prometheus replicas.

Reply:
Good catch, fixed.